lib/active_recall/algorithms/soft_leitner_system.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module ActiveRecall
+  class SoftLeitnerSystem
+    DELAYS = [3, 7, 14, 30, 60, 120, 240].freeze
+
+    def self.right(box:, times_right:, times_wrong:, current_time: Time.current)
+      new(
+        box: box,
+        current_time: current_time,
+        times_right: times_right,
+        times_wrong: times_wrong
+      ).right
+    end
+
+    def self.wrong(box:, times_right:, times_wrong:, current_time: Time.current)
+      new(
+        box: box,
+        current_time: current_time,
+        times_right: times_right,
+        times_wrong: times_wrong
+      ).wrong
+    end
+
+    def initialize(box:, times_right:, times_wrong:, current_time: Time.current)
+      @box = box
+      @current_time = current_time
+      @times_right = times_right
+      @times_wrong = times_wrong
+    end
+
+    def right
+      self.box = [box + 1, DELAYS.count].min
+
+      {
+        box: box,
+        times_right: times_right + 1,
+        times_wrong: times_wrong,
+        last_reviewed: current_time,
+        next_review: next_review
+      }
+    end
+
+    def wrong
+      self.box = [box - 1, 0].max
```

Comment:
I agree that the original algorithm is far too harsh, but maybe going back only a single box is too lenient. I'm could easily be convinced of -2 or -3, I think.

Reply:
In my implementation, I used this:

`next_review: current_time + rand(1..14).days`

The idea is that setting the card to be reviewed on a random day within 2 weeks means users who guess lots of cards wrong won't find themselves overwhelmed by a huge number of cards to review on their next review session.

I think your implementation with box - 2 would be a good middle ground. My only concern is that this could cause lots of items to be stuck in box 0. Not sure how much of an issue it would be in real use case scenario though. I ran a few simulations with slightly different calculations but I couldn't come up with anything better so I think we can go with this version for now. We can always finetune it based on user feedback.

Reply:
> The idea is that setting the card to be reviewed on a random day within 2 weeks means users who guess lots of cards wrong won't find themselves overwhelmed by a huge number of cards to review on their next review session. This essentially mean that while items would only go down one box, they would still be reviewed within the next 2 weeks. A possibility would be to use this randomized review time together with box - 2 so the demotion happens more rapidly.

This is a great point and I agree preventing a huge buildup of due cards is a bad user experience, but I also think handling that issue at the level of this library would dilute its focus somewhat versus doing so in the application logic. My intent is for this library to merely handle review scheduling for any active record model and let the user of the library determine what to do with that information.

One of the apps that uses this has a concept of a "daily workout":
```ruby
create_table "daily_workouts", force: :cascade do |t|
    t.time "completed_at"
    t.date "date"
    t.bigint "user_id", null: false
    t.datetime "created_at", null: false
    t.datetime "updated_at", null: false
    t.integer "repetition_goal", default: 25
    t.integer "repetitions", default: 0
    t.index ["user_id"], name: "index_daily_workouts_on_user_id"
  end
```
Basically, there is a goal number of reviews per day, and we count the number of exercise repetitions. Once their reps exceeds that goal, then their daily workout is considered "completed" (even if `ActiveRecall` says there are still some cards remaining).

> I think your implementation with box - 2 could work well too. My only concern is that this could cause lots of items to be stuck in box 0. Not sure how much of an issue it would be in real use case scenario though.

>I honestly don't know which version is better as I think they both have their merits and could both work well so I would suggest going with whichever you prefer and we can later finetune it based on users feedback.

I'll leave it as is for now, but we can easily change later, or even make it configurable just like the SRS algorithm. 👍🏼